Refuels

JPS09 opened the pull request:
closes #20 #21 #22

JPS09 committed:
Added basic crud methods and index view

JPS09 committed:
Added very basic layout to insure working back-end logic

JPS09 committed:
Added working form to add refuel

JPS09 committed:
Added working delete

JPS09 committed:
Added validations

JPS09 committed:
Added notice when creation and delete

JPS09 committed:
Added error message and prevented negatives inputs

JPS09 committed:
Ajout d'un lien vers la fiche technique ainsi que d'un retour à l'index dans la création

JPS09 committed:
Added basic styling on the forms

JPS09 committed:
Further testing of form css

JPS09 committed:
Cleaning old simple form to remove bootstrap

JPS09 committed:
Added clean simple form without bootstrap

JPS09 committed:
Removed custom class

sbriclot commented on app/controllers/refuels_controller.rb:
```diff
@@ -1,2 +1,37 @@
+  before_action :find_vehicle, only: %i[index create new]
+
+  def index
+    @refuels = Refuel.where(vehicle_id: params[:vehicle_id])
```
Sort the index in descending order (last refuel on top of the list)

sbriclot commented on app/models/refuel.rb:
```diff
@@ -1,4 +1,11 @@
   belongs_to :vehicle
+
+  validates :date, presence: true
+  validates :km, presence: true, numericality: { greater_than: 0 }
+  validates :quantity, presence: true, numericality: { greater_than: 0 }
+  validates :price, presence: true, numericality: { greater_than: 0 }
```
Float scales must be fixed (2 decimals for quantity and price, 1 decimal for km).
You can do that in the model validations, or in the db with a migration file, feel free to use your favorite method.

JPS09 replied:
Do you think it would be appropriate to change all floats in the DB into 2 steps decimals?
I mean, we also have price, km columns in other table beside this one, do we leave them be or do we change them all?

sbriclot replied:
Price has to be adapted in INVOICES, but it's not necessary to adapt km in other tables.
For REFUELS, we use the daily trip odometer which is displayed with 1 decimal on the dashborad.
But in other tables, we use the total odometer which is an integer.

JPS09 replied:
Ok, I tried to implement this by migrating but the current stored values are too big so I tried some SQL to trunk their decimal values in the migrations, thanks to https://www.techonthenet.com/postgresql/functions/trunc.php but it's not working when I add the precision, can you tell me why ?
![image](https://user-images.githubusercontent.com/58084593/107939522-25c70480-6f87-11eb-87bf-9d256dbe8e65.png)

It does what I want (kinda) without this argument
SQL : SELECT trunc(price) FROM refuels
result:
![image](https://user-images.githubusercontent.com/58084593/107939682-5dce4780-6f87-11eb-9410-2f26f717da5d.png)

sbriclot replied:
I don't know why TRUNC doesn't work with decimals, I have the same issue here.
I prefer to use ROUND, you can execute the follow ActiveRecord query in the Rails console :
`Refuel.update_all("price = ROUND(price::numeric,2)");`

Or you can execute a `rails db:seed` first, to reset the table REFUELS with rounded values.

JPS09 replied:
I wanted to do something different, something that could be effective in a bigger scale application/collaboration,  so that you the team only need to run the migration to update the database. I don't know if we can run activeRecord query inside a migration, I will try your suggestion as soon as I can. Thank you for trying it out on your side

sbriclot commented on app/views/refuels/_form.html.erb:
```diff
@@ -0,0 +1,12 @@
+		<%= f.input :km, label: 'Kilométrage au moment du plein', input_html: { min:0 }, error:'Veuillez renseigner un kilométrage qui soit supérieur à 0' %>
+		<%= f.input :quantity,label: 'Quantité de carburant', input_html: { min:0 }, error: 'Veuillez renseigner une quantité qui soit supérieure à 0'%>
+		<%= f.input :price, label: 'Prix', input_html: { min:0 }, error: 'Veuillez renseigner un prix qui soit supérieur à 0' %>
+		<%= f.input :fuel_id, collection: Fuel.all, label: 'Carburant utilisé', prompt:'Sélectionner un carburant', error: 'Veuillez sélectionner un carburant', input_html:{class:'select'}  %>
```
```suggestion
		<%= f.input :fuel_id, collection: VehicleFuel.all, label: 'Carburant utilisé', prompt:'Sélectionner un carburant', error: 'Veuillez sélectionner un carburant', input_html:{class:'select'}  %>
```
Limit the collection to the fuels allowed for the vehicle.
```suggestion
		<%= f.input :fuel_id, collection: Fuel.all, label: 'Carburant utilisé', prompt:'Sélectionner un carburant', error: 'Veuillez sélectionner un carburant', input_html:{class:'select'}  %>
```

sbriclot reviewed (changes requested):
Last change to do to fix my mistake in the seed file, please change line 134 to round the prices :
refuel_price = (refuel_quantity * Random.new.rand(1.10..1.35)).round(2)
(should work)

JPS09 committed:
Sorted refuels in index by descending order and created a method to find fuels of one vehicle to use in form

JPS09 committed:
Fixed seed files according to suggestion

JPS09 committed:
Created migration to change particular columns to be decimals and allowing to rollback

JPS09 committed:
Fixed the migration

JPS09 committed:
Removed migration since need to truncate data which can lead to lost data and irreversability

JPS09 committed:
restored migration and added sql which is not working yet so commented

JPS09 committed:
Added sql

sbriclot commented on app/controllers/refuels_controller.rb:
```diff
@@ -1,2 +1,46 @@
+
+  def find_fuel
+    fuel_ids = []
+    VehicleFuel.find_each do |fuel|
```
```suggestion
    VehicleFuel.where(vehicle_id: params[:vehicle_id]).find_each do |fuel|
```
Add a WHERE clause to select allowed fuel(s) for the current vehicle only.

JPS09 commented:
TODO : Add a link to vehicle show

JPS09 committed:
Updated SQL queries in migration

JPS09 committed:
Working migrations

JPS09 committed:
Added conditions on vehicle fuel

JPS09 committed:
Added remote true to delete

JPS09 committed:
init chart js, WIP

JPS09 committed:
Added comment to prevent error for now

JPS09 committed:
Added instance variable to store dates and price for graph

JPS09 committed:
Working placeholder chart

JPS09 committed:
Chart styling

JPS09 committed:
Added data from DB into the chart

JPS09 committed:
Added dates to graph from the DB

JPS09 committed:
Added a WIP average chart to show the average spending for each month

JPS09 committed:
Pseudo code

JPS09 committed:
added titles to describe chart purpose

JPS09 committed:
Fixed missing color on first chart

JPS09 committed:
Fixed data shifting

JPS09 committed:
Switched data to improve front

JPS09 committed:
Added a link to refuel section in the vehicle show

JPS09 committed:
Text change on refuel details

sbriclot commented:
I love the first chart 👍, good job.

JPS09 commented:
> 
> 
> I love the first chart 👍, good job.

Happy that it suits you, I will try to work on the other one asap :)

JPS09 committed:
Cleaning index method by creating a private method for chart

JPS09 committed:
Started working on twelve month history refuel chart

JPS09 committed:
Fetching data depending on the refuel date

JPS09 committed:
Comments on what step to take next to do a monthly filter in order on refuels

JPS09 committed:
Started working on the dynamic labels

sbriclot commented on app/controllers/refuels_controller.rb:
```diff
@@ -1,2 +1,81 @@
+  before_action :find_fuels, only: %i[new create]
+
+  def index
+    @refuels = Refuel.where(vehicle_id: params[:vehicle_id]).order(created_at: :desc)
```
```suggestion
    @refuels = Refuel.where(vehicle_id: params[:vehicle_id]).order(:created_at).limit(10)
```
Because we don't need the full index in the final version, you can directly select the 10 last refuels here, and change the order. Then you can remove the commands **take(10).reverse** (3x) in the HTML file.
This way will be faster, you avoid to load 3 full arrays in the method **last_ten_refuels_info**, you only load the useful set of data.

JPS09 replied:
I understand your concerns but I wanted to try to implement a slider input to change how many refuels are displayed in the graph, that could be a way to have fun with js, what do you think?

sbriclot replied:
The slider is a good idea.
Just use a step of 5 or 10, and add an event on mouseup only, to limit the reloads.
Do you have an idea to redraw the graph without refreshing the full page ?
For me, the datasets are popuated when the page is loaded. So, if you change the slider value, you have to reload the page to update the datasets.
In this case, each refresh will load the full index and populate the 3 full arays. So it's better to send the slider value to the controller to limit the records returned by the controller.

But maybe I'm wrong and you can redraw the graph without refreshing the full page.

Last thing, I made a mistake with my suggestion (if you need it), it's not **limit(n)** but **last(n)** to retrun the nth last records.

sbriclot commented:
For the 12 months stat, I have 2 suggestions for you, one using ActiveRecord, the second using SQL.
**ActiveRecord**
The query should be

sbriclot commented:
For the 12 months stat, I have 2 suggestions for you, one using ActiveRecord, the second using SQL.
**ActiveRecord**
The query could be :
```
@refuel_history = Refuel.where("date >= current_date - 365")
                        .order("DATE_TRUNC('month', date)")
                        .group("DATE_TRUNC('month', date)")
                        .average(:price)
```
This query returns a hash, and i don't understand how to attach alias to the result.
So, the associated HTML code for this query could be :
```
<% @refuel_history.each do |r| %>
  <p><%= r[0].strftime("%m/%Y") %></p>
  <p><%= r[1].round(2) %></p>
<% end %>
```

**SQL**
The code in the controller could be :
```
    sql = "select date_trunc('month', date) as month, round(avg(price), 2) as avg_price
           from   refuels
           where  date > current_date - 365
           group by month
           order by month"
    @refuel_history = ActiveRecord::Base.connection.exec_query(sql)
```
With this version, I can use the alias (as string), and the average price is already roounded, sot he HTML code could be :
```
<% @refuel_history.each do |r| %>
  <p><%= r["month"].strftime("%m/%Y") %></p>
  <p><%= r["avg_price"] %></p>
<% end %>
```

The result is the same, use or adapt your favorite one.
For the ActiveRecord version, maybe you can try to use alias, I made some tests but I always have an empty result (even if the SQL code generated and displayed in the rails console works fine in DBeaver).

JPS09 commented:
> 
> 
> For the 12 months stat, I have 2 suggestions for you, one using ActiveRecord, the second using SQL.
> **ActiveRecord**
> The query could be :
> 
> ```
> @refuel_history = Refuel.where("date >= current_date - 365")
>                         .order("DATE_TRUNC('month', date)")
>                         .group("DATE_TRUNC('month', date)")
>                         .average(:price)
> ```
> 
> This query returns a hash, and i don't understand how to attach alias to the result.
> So, the associated HTML code for this query could be :
> 
> ```
> <% @refuel_history.each do |r| %>
>   <p><%= r[0].strftime("%m/%Y") %></p>
>   <p><%= r[1].round(2) %></p>
> <% end %>
> ```
> 
> **SQL**
> The code in the controller could be :
> 
> ```
>     sql = "select date_trunc('month', date) as month, round(avg(price), 2) as avg_price
>            from   refuels
>            where  date > current_date - 365
>            group by month
>            order by month"
>     @refuel_history = ActiveRecord::Base.connection.exec_query(sql)
> ```
> 
> With this version, I can use the alias (as string), and the average price is already roounded, sot he HTML code could be :
> 
> ```
> <% @refuel_history.each do |r| %>
>   <p><%= r["month"].strftime("%m/%Y") %></p>
>   <p><%= r["avg_price"] %></p>
> <% end %>
> ```
> 
> The result is the same, use or adapt your favorite one.
> For the ActiveRecord version, maybe you can try to use alias, I made some tests but I always have an empty result (even if the SQL code generated and displayed in the rails console works fine in DBeaver).

Your suggestion reminded me that it could be done from within the DB queries, I totally forgot about those! I will try to tinker with that to add aliases :) Thank you !

JPS09 committed:
Fixed average query

JPS09 committed:
Adjusted labels

JPS09 committed:
Removed Refuels placeholder

JPS09 committed:
Using the price color from above chart

JPS09 committed:
Added a slider

JPS09 committed:
Added dynamic title number

JPS09 committed:
Trying to handle load bug

JPS09 committed:
Removed useless function

JPS09 committed:
Comment on next step to do

JPS09 committed:
Remove data function beta

JPS09 committed:
Added dynamic step value based on input element

JPS09 committed:
Adjusted css class name and added ouput slider value update when slider value changes

JPS09 committed:
File renaming for new purpose

JPS09 committed:
Started styling range slider

JPS09 committed:
Added opacity and transition on hover

JPS09 committed:
First pass at styling the slider

JPS09 committed:
Opacity customisation

JPS09 committed:
Adjusted opacity for slider body

JPS09 committed:
fixed output displaying

JPS09 committed:
Little To dos and updates

JPS09 committed:
Added units to tooltips for first chart

JPS09 committed:
Added units for second charts

JPS09 committed:
Make the second chart start at zero for now

JPS09 committed:
Working slider for first chart

JPS09 committed:
Added label to slider

JPS09 committed:
Centered slider

JPS09 committed:
Added date picker

JPS09 committed:
Little refacto

JPS09 commented:
Last updates : 

- Added units on tooltip hovering for each charts
- Found a way to make the slider influence the chart correctly.
- Added a visual output of the slider value
- Made the second chart's legend start at zero
- Added flatpickr for this CRUD form

Can you (@sbriclot) check if it's working properly on your side and tell me if it can be merged?

JPS09 committed:
Changed let to const for first graph

JPS09 committed:
Fixing data shifting

JPS09 committed:
Revert datapickr implementation to avoid overloading PR

sbriclot commented on app/javascript/components/refuels/outputSliderValue.js:
```diff
@@ -0,0 +1,15 @@
+    titleSpan.textContent = inputSlider.value;
+  };
+  if (inputSlider) {
+    inputSlider.addEventListener("load", updateValue);
```
```suggestion
    window.addEventListener("load", updateValue);
```

sbriclot replied:
Using **window** instead of **inputSlider** here updates the **title** and **outputElement** values to the previous slider value on page refresh.

sbriclot commented on app/javascript/plugins/_init_chartjs.js:
```diff
@@ -0,0 +1,126 @@
+  if (ctxRefuels) {
+    const price = JSON.parse(ctxRefuels.dataset.price);
+    const quantity = JSON.parse(ctxRefuels.dataset.quantity);
+    const dates = JSON.parse(ctxRefuels.dataset.dates);
+    const refuelChartCanvas = new Chart(ctxRefuels, {
```
```suggestion
    const dates = JSON.parse(ctxRefuels.dataset.dates);
    const sliderInput = document.querySelector("#refuel-slider");
    const refuelChartCanvas = new Chart(ctxRefuels, {
```

sbriclot commented on app/javascript/plugins/_init_chartjs.js:
```diff
@@ -0,0 +1,126 @@
+    const refuelChartCanvas = new Chart(ctxRefuels, {
+      type: "bar",
+      data: {
+        labels: dates,
```
```suggestion
        labels: dates.slice(-sliderInput.value),
```

sbriclot commented on app/javascript/plugins/_init_chartjs.js:
```diff
@@ -0,0 +1,126 @@
+        datasets: [
+          {
+            label: "Quantité de carburant",
+            data: quantity,
```
```suggestion
            data: quantity.slice(-sliderInput.value),
```

sbriclot commented on app/javascript/plugins/_init_chartjs.js:
```diff
@@ -0,0 +1,126 @@
+          },
+          {
+            label: "Prix du plein",
+            data: price,
```
```suggestion
            data: price.slice(-sliderInput.value),
```

sbriclot reviewed (changes requested):
The change below displays 10 bars only (instead of the full dataset) on page first load.

JPS09 committed:
Applied suggestions

JPS09 commented:
> 
> 
> The change below displays 10 bars only (instead of the full dataset) on page first load.

Thank you for your advice, I applied them. Can you check one last time before merging?

sbriclot commented:
Good for me, you can merge this branch now.

JPS09 committed:
Merge branch 'master' into Refuels
